polkadot/node/network/availability-recovery/src/task/strategy/systematic.rs

```diff
@@ -0,0 +1,350 @@
+// Copyright (C) Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::{
+	futures_undead::FuturesUndead,
+	task::{
+		strategy::{
+			do_post_recovery_check, is_unavailable, OngoingRequests, N_PARALLEL,
+			SYSTEMATIC_CHUNKS_REQ_RETRY_LIMIT,
+		},
+		RecoveryParams, RecoveryStrategy, State,
+	},
+	ErasureTask, LOG_TARGET,
+};
+
+use polkadot_node_primitives::AvailableData;
+use polkadot_node_subsystem::{overseer, RecoveryError};
+use polkadot_primitives::{ChunkIndex, ValidatorIndex};
+
+use std::collections::VecDeque;
+
+/// Parameters needed for fetching systematic chunks.
+pub struct FetchSystematicChunksParams {
+	/// Validators that hold the systematic chunks.
+	pub validators: Vec<(ChunkIndex, ValidatorIndex)>,
+	/// Validators in the backing group, to be used as a backup for requesting systematic chunks.
+	pub backers: Vec<ValidatorIndex>,
```

Comment:
I doubt this is useful if we first try to fetch from backers and fail. This only makes sense if we attempt this strategy first.

Reply:
with the strategy that's going to be used in production, we'll only try from backers first for small pov sizes.

if it fails, then indeed it's arguable whether it's useful or not. But we'll only use the backers for systematic chunks if the systematic validators didn't supply them (so it's a corner case after a corner case). And in the meantime, maybe a backer came back online.

So I think it could still be useful, but it's not the case I was optimising for indeed

Reply:
> with the strategy that's going to be used in production, we'll only try from backers first for small pov sizes.
> 

I've done some tests with gluttons a while back on Versi and it turns out that we can move that threshold for small PoV to smth like 1/2 max size. I would argue that we could always fetch from backers unless the backing group is 2 or 3.

> if it fails, then indeed it's arguable whether it's useful or not. But we'll only use the backers for systematic chunks if the systematic validators didn't supply them (so it's a corner case after a corner case). And in the meantime, maybe a backer came back online.
> 
> So I think it could still be useful, but it's not the case I was optimising for indeed

Sure, this is reasonable.

Reply:
> I've done some tests with gluttons a while back on Versi and it turns out that we can move that threshold for small PoV to smth like 1/2 max size. I would argue that we could always fetch from backers unless the backing group is 2 or 3.

could be

an interesting thing to note is that prior to this PR, the POV size estimate was being done using the regular recovery threshold, not the systematic recovery threshold, which gives much less accurate results.

Systematic threshold can be up to half of the regular threshold (because it's the next lower power of two). So, if we multiply the chunk size by the systematic threshold instead, we get a much more accurate estimate of the plain text size.

The bottom line is that in the calculated estimate so far, I think we have been overestimating by up to 100%. This could have given us some false results when testing, because we were pessimistic about the POV size and put less pressure on the backers than intended.

In any case, it's an interesting thing to test, orthogonal of this PR